# India Food Prices Project

## Setup

In [1]:
import pandas as pd
import numpy as np
df_food = pd.read_csv('wfp_food_prices_ind.csv')
df_food.head()

/var/folders/sx/wtgz65jx1vqg_v5k12yf0wfr0000gn/T/ipykernel_74588/3192963853.py:3: DtypeWarning: Columns (4,5,12,13) have mixed types. Specify dtype option on import or set low_memory=False.
  df_food = pd.read_csv('wfp_food_prices_ind.csv')


date      admin1      admin2            market   latitude  longitude  \
0       #date  #adm1+name  #adm2+name  #loc+market+name   #geo+lat   #geo+lon   
1  1994-01-15       Delhi       Delhi             Delhi  28.666667  77.216667   
2  1994-01-15       Delhi       Delhi             Delhi  28.666667  77.216667   
3  1994-01-15       Delhi       Delhi             Delhi  28.666667  77.216667   
4  1994-01-15       Delhi       Delhi             Delhi  28.666667  77.216667   

             category      commodity        unit         priceflag  \
0          #item+type     #item+name  #item+unit  #item+price+flag   
1  cereals and tubers           Rice          KG            actual   
2  cereals and tubers          Wheat          KG            actual   
3  miscellaneous food          Sugar          KG            actual   
4        oil and fats  Oil (mustard)          KG            actual   

          pricetype   currency   price    usdprice  
0  #item+price+type  #currency  #value  #value+usd  
1            Retail        INR     8.0      0.2545  
2            Retail        INR     5.0       0.159  
3            Retail        INR    13.5      0.4294  
4            Retail        INR    31.0       0.986

## Cleaning the dataset

In [2]:
df_food.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 172018 entries, 0 to 172017
Data columns (total 14 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   date       172018 non-null  object
 1   admin1     171229 non-null  object
 2   admin2     171229 non-null  object
 3   market     172018 non-null  object
 4   latitude   171229 non-null  object
 5   longitude  171229 non-null  object
 6   category   172018 non-null  object
 7   commodity  172018 non-null  object
 8   unit       172018 non-null  object
 9   priceflag  172018 non-null  object
 10  pricetype  172018 non-null  object
 11  currency   172018 non-null  object
 12  price      172018 non-null  object
 13  usdprice   172018 non-null  object
dtypes: object(14)
memory usage: 18.4+ MB


In [3]:
df_food = df_food.loc[1:, :]
df_food.head()

date   admin1     admin2     market   latitude  longitude  \
1  1994-01-15    Delhi      Delhi      Delhi  28.666667  77.216667   
2  1994-01-15    Delhi      Delhi      Delhi  28.666667  77.216667   
3  1994-01-15    Delhi      Delhi      Delhi  28.666667  77.216667   
4  1994-01-15    Delhi      Delhi      Delhi  28.666667  77.216667   
5  1994-01-15  Gujarat  Ahmadabad  Ahmedabad  23.033333  72.616667   

             category      commodity unit priceflag pricetype currency price  \
1  cereals and tubers           Rice   KG    actual    Retail      INR   8.0   
2  cereals and tubers          Wheat   KG    actual    Retail      INR   5.0   
3  miscellaneous food          Sugar   KG    actual    Retail      INR  13.5   
4        oil and fats  Oil (mustard)   KG    actual    Retail      INR  31.0   
5  cereals and tubers           Rice   KG    actual    Retail      INR   6.8   

  usdprice  
1   0.2545  
2    0.159  
3   0.4294  
4    0.986  
5   0.2163

In [4]:
df_food['priceflag'].value_counts()

actual    172017
Name: priceflag, dtype: int64

In [5]:
df_food['currency'].value_counts()

INR    172017
Name: currency, dtype: int64

Some basic problems with the dataset so far:
1. Redundant columns like `priceflag` and `currency` that contain only one unique value. Need to be removed.
2. All columns have `object` dtype, need to be corrected to most appropriate dtype.
3. Some columns would be more useful if renamed For example: `admin1`, `admin2`, `price`.
4. We are only interested in price in INR, the domestic currency, so the `usdprice` column becomes redundant.
5. `latitude` and `longitude` are not relevant to our analysis.
6. Dates are not uniform across states--for some states the earliest records are 1994, for some they are 2012.

In [6]:
#Correcting dtypes
df_food['admin1'] = df_food['admin1'].astype('category')
df_food['admin2'] = df_food['admin2'].astype('category')
df_food['market'] = df_food['market'].astype('category')
df_food['latitude'] = df_food['latitude'].astype('float64')
df_food['longitude'] = df_food['longitude'].astype('float64')
df_food['category'] = df_food['category'].astype('category')
df_food['commodity'] = df_food['commodity'].astype('category')
df_food['unit'] = df_food['unit'].astype('category')
df_food['priceflag'] = df_food['priceflag'].astype('category')
df_food['pricetype'] = df_food['pricetype'].astype('category')
df_food['price'] = df_food['price'].astype('float64')
df_food['usdprice'] = df_food['usdprice'].astype('float64')

In [7]:
#drop 'currency', 'priceflag', 'usdprice'
df_food = df_food.drop(['priceflag', 'currency', 'usdprice', 'latitude', 'longitude'], axis=1)

#rename 'admin1', 'admin2', and 'price'
df_food = df_food.rename({'price':'inrprice', 'admin1':'state', 'admin2':'district'}, axis=1)

In [8]:
df_food.sort_values(by = 'date')

date        state        district     market  \
1       1994-01-15        Delhi           Delhi      Delhi   
33      1994-01-15  West Bengal         Kolkata    Kolkata   
32      1994-01-15  West Bengal         Kolkata    Kolkata   
31      1994-01-15  West Bengal         Kolkata    Kolkata   
30      1994-01-15      Tripura    West Tripura   Agartala   
...            ...          ...             ...        ...   
169833  2022-07-15    Jharkhand       Lohardaga  Lohardaga   
169834  2022-07-15    Jharkhand       Lohardaga  Lohardaga   
169835  2022-07-15    Jharkhand       Lohardaga  Lohardaga   
169825  2022-07-15    Jharkhand       Lohardaga  Lohardaga   
172017  2022-07-15  West Bengal  West Midnapore  Kharagpur   

                     category            commodity unit pricetype  inrprice  
1          cereals and tubers                 Rice   KG    Retail      8.00  
33               oil and fats        Oil (mustard)   KG    Retail     31.00  
32         miscellaneous food                Sugar   KG    Retail     13.40  
31         cereals and tubers                 Rice   KG    Retail      6.75  
30               oil and fats        Oil (mustard)   KG    Retail     41.00  
...                       ...                  ...  ...       ...       ...  
169833        pulses and nuts              Lentils   KG    Retail     98.07  
169834        pulses and nuts      Lentils (masur)   KG    Retail     90.30  
169835        pulses and nuts      Lentils (moong)   KG    Retail     98.07  
169825     miscellaneous food  Sugar (jaggery/gur)   KG    Retail     43.20  
172017  vegetables and fruits             Tomatoes   KG    Retail     53.70  

[172017 rows x 9 columns]

In [9]:
#Checking for null values
print(df_food['state'].isna().sum())
df_food['district'].isna().sum()

789


789

In [10]:
#Dropping null values from our df (only missing values are in 'state' and 'district' columns,)
df_food = df_food[pd.notnull(df_food['state']) & pd.notnull(df_food['district'])]
df_food.isna().sum().sum()

0

In [11]:
df_food[df_food['unit']=='L'].sort_values(by='date')

date           state         district      market  \
10845   2009-01-15           Bihar            Patna       Patna   
10855   2009-01-15           Delhi            Delhi       Delhi   
10863   2009-01-15         Gujarat        Ahmadabad   Ahmedabad   
10871   2009-01-15       Karnataka  Bangalore Urban   Bengaluru   
10877   2009-01-15  Madhya Pradesh           Bhopal      Bhopal   
...            ...             ...              ...         ...   
169843  2022-07-15       Jharkhand  Purba Singhbhum  Jamshedpur   
169864  2022-07-15       Jharkhand           Ranchi      Ranchi   
169914  2022-07-15       Karnataka  Bangalore Urban   Bengaluru   
170367  2022-07-15  Madhya Pradesh         Jabalpur    Jabalpur   
172001  2022-07-15     West Bengal   West Midnapore   Kharagpur   

              category           commodity unit pricetype  inrprice  
10845   milk and dairy                Milk    L    Retail     20.00  
10855   milk and dairy                Milk    L    Retail     21.00  
10863   milk and dairy                Milk    L    Retail     24.20  
10871   milk and dairy                Milk    L    Retail     16.00  
10877   milk and dairy                Milk    L    Retail     22.05  
...                ...                 ...  ...       ...       ...  
169843  milk and dairy  Milk (pasteurized)    L    Retail     48.00  
169864  milk and dairy  Milk (pasteurized)    L    Retail     52.00  
169914  milk and dairy  Milk (pasteurized)    L    Retail     41.84  
170367  milk and dairy  Milk (pasteurized)    L    Retail     56.59  
172001  milk and dairy  Milk (pasteurized)    L    Retail     45.00  

[9009 rows x 9 columns]

In [12]:
df_food[df_food['unit']=='KG'].sort_values(by='date')

date        state        district     market  \
1       1994-01-15        Delhi           Delhi      Delhi   
33      1994-01-15  West Bengal         Kolkata    Kolkata   
32      1994-01-15  West Bengal         Kolkata    Kolkata   
31      1994-01-15  West Bengal         Kolkata    Kolkata   
30      1994-01-15      Tripura    West Tripura   Agartala   
...            ...          ...             ...        ...   
169834  2022-07-15    Jharkhand       Lohardaga  Lohardaga   
169835  2022-07-15    Jharkhand       Lohardaga  Lohardaga   
169836  2022-07-15    Jharkhand       Lohardaga  Lohardaga   
169813  2022-07-15    Jharkhand           Gumla      Gumla   
172017  2022-07-15  West Bengal  West Midnapore  Kharagpur   

                     category        commodity unit pricetype  inrprice  
1          cereals and tubers             Rice   KG    Retail      8.00  
33               oil and fats    Oil (mustard)   KG    Retail     31.00  
32         miscellaneous food            Sugar   KG    Retail     13.40  
31         cereals and tubers             Rice   KG    Retail      6.75  
30               oil and fats    Oil (mustard)   KG    Retail     41.00  
...                       ...              ...  ...       ...       ...  
169834        pulses and nuts  Lentils (masur)   KG    Retail     90.30  
169835        pulses and nuts  Lentils (moong)   KG    Retail     98.07  
169836        pulses and nuts   Lentils (urad)   KG    Retail     98.07  
169813        pulses and nuts  Lentils (masur)   KG    Retail     93.79  
172017  vegetables and fruits         Tomatoes   KG    Retail     53.70  

[159942 rows x 9 columns]

In [13]:
bool_mask1 = df_food['unit'] == '100 KG'
bool_mask2 = df_food['pricetype'] == 'Retail'
df_food[bool_mask1 & bool_mask2]

Empty DataFrame
Columns: [date, state, district, market, category, commodity, unit, pricetype, inrprice]
Index: []

In [14]:
bool_mask3 = df_food['unit'] == 'KG'
df_food[bool_mask3 & ~bool_mask2]

Empty DataFrame
Columns: [date, state, district, market, category, commodity, unit, pricetype, inrprice]
Index: []

In [15]:
bool_mask4 = df_food['unit'] == 'L'
df_food[bool_mask4 & ~bool_mask2] #36

Empty DataFrame
Columns: [date, state, district, market, category, commodity, unit, pricetype, inrprice]
Index: []

There are three different measures of weight, for different products and over different time periods. This complicates things, and we should either standardise or drop those values that we don't need.

We can see that `100 KG`  corresponds exclusively to pricetype `wholesale`. Meanwhile, both `L` and `KG` correspond to `retail` pricetype. This information will be useful when wrangling the data.

## Data Wrangling

Having done some initial cleaning, we can now wrangle the dataset to prepare it for further analysis.

Questions we can ask:
1. What states had highest/lowest food inflation since 1994?
2. What states had highest/lowest food inflation in the post-pandemic opening-up?
3. The above two questions, but for categories and commodities of food instead of states
4. Answering all above questions separately for wholesale and retail inflation

To answer these questions, need to first complete some tasks:
1. Create separate dfs for retail and wholesale calculations
2. Creating a new df with yearly indices of inflation: need to create a basket of goods to calculate price change with
3. Editing the df such that we have a uniform timeframe across states/UTs to work with

In [16]:
df_food['state'].value_counts()

Uttar Pradesh          13173
Bihar                  11337
Madhya Pradesh         11320
Tamil Nadu             10626
Orissa                 10382
Maharashtra            10077
Karnataka               8541
Rajasthan               8397
Haryana                 7985
Kerala                  7637
Gujarat                 7257
West Bengal             7143
Punjab                  6490
Himachal Pradesh        6083
Telangana               5225
Delhi                   4345
Jharkhand               4285
Meghalaya               3556
Andhra Pradesh          3549
Uttarakhand             3530
Assam                   2588
Andaman and Nicobar     2380
Tripura                 2271
Chandigarh              2241
Nagaland                2190
Puducherry              2152
Chhattisgarh            2069
Goa                     1966
Mizoram                 1808
Sikkim                   400
Manipur                  225
Name: state, dtype: int64

Not all states have data up until 1994. We need to find a better starting point that encapsulates more (if not all) states. This has the added benefit of making the analysis more contemporary. For this, it would help to isolate `year` from the `date`.

In [17]:
df_food['year'] = df_food['date'].str.split(pat='-', expand=True)[0]

In [18]:
df_food = df_food.drop(['date'], axis=1)

In [19]:
df_food.head()

state   district     market            category      commodity unit  \
1    Delhi      Delhi      Delhi  cereals and tubers           Rice   KG   
2    Delhi      Delhi      Delhi  cereals and tubers          Wheat   KG   
3    Delhi      Delhi      Delhi  miscellaneous food          Sugar   KG   
4    Delhi      Delhi      Delhi        oil and fats  Oil (mustard)   KG   
5  Gujarat  Ahmadabad  Ahmedabad  cereals and tubers           Rice   KG   

  pricetype  inrprice  year  
1    Retail       8.0  1994  
2    Retail       5.0  1994  
3    Retail      13.5  1994  
4    Retail      31.0  1994  
5    Retail       6.8  1994

Apart from Chhattisgarh, Sikkim, Manipur, we have data on all states/Union Territories from 2012 onwards. For convenience, let us use 2014 as a reference point for analysis, coinciding with the beginning of the present government's first tenure.

To further simplify the data for this purpose, let us exclude the above three states from the dataframe.

In [20]:
cg_filter = df_food['state'] == 'Chhattisgarh'
sk_filter = df_food['state'] == 'Sikkim'
mn_filter = df_food['state'] == 'Manipur'
df_food = df_food[~cg_filter & ~sk_filter & ~mn_filter]
df_food['state'].value_counts()

Uttar Pradesh          13173
Bihar                  11337
Madhya Pradesh         11320
Tamil Nadu             10626
Orissa                 10382
Maharashtra            10077
Karnataka               8541
Rajasthan               8397
Haryana                 7985
Kerala                  7637
Gujarat                 7257
West Bengal             7143
Punjab                  6490
Himachal Pradesh        6083
Telangana               5225
Delhi                   4345
Jharkhand               4285
Meghalaya               3556
Andhra Pradesh          3549
Uttarakhand             3530
Assam                   2588
Andaman and Nicobar     2380
Tripura                 2271
Chandigarh              2241
Nagaland                2190
Puducherry              2152
Goa                     1966
Mizoram                 1808
Chhattisgarh               0
Sikkim                     0
Manipur                    0
Name: state, dtype: int64

In [21]:
#Removing pre-2014 data
df_food['year'] = df_food['year'].astype('int64')
df_food = df_food[df_food['year'] >= 2014]
df_food['year'].describe()

count    139199.000000
mean       2018.852657
std           2.591697
min        2014.000000
25%        2017.000000
50%        2020.000000
75%        2021.000000
max        2022.000000
Name: year, dtype: float64

Taking stock: so far we have cleaned our dataset, and among other things, removed null values and made our timeframe uniform across states. Now our entire dataset represents prices of commodities across states (excluding Chhatisgarh, Sikkim, and Manipur) in the timeframe 2014-2022.

Now we proceed to separate our analyses for retail and wholesale inflation.

In [22]:
#New dfs for retail and wholesale inflation
df_retail = df_food[df_food['pricetype'] == 'Retail']
df_wholesale = df_food[df_food['pricetype'] == 'Wholesale']

In [23]:
df_retail['state'].value_counts()

Uttar Pradesh          11239
Madhya Pradesh         10107
Bihar                   9051
Orissa                  8828
Tamil Nadu              8037
Karnataka               7506
Haryana                 7338
Maharashtra             7055
Rajasthan               6875
Kerala                  6376
West Bengal             5961
Gujarat                 5838
Punjab                  5553
Himachal Pradesh        5031
Telangana               4202
Jharkhand               3988
Andhra Pradesh          3200
Uttarakhand             3184
Meghalaya               2703
Delhi                   2121
Andaman and Nicobar     2106
Goa                     1937
Chandigarh              1883
Puducherry              1875
Nagaland                1858
Assam                   1782
Tripura                 1596
Mizoram                 1149
Chhattisgarh               0
Sikkim                     0
Manipur                    0
Name: state, dtype: int64

In [24]:
df_wholesale['state'].value_counts()

Maharashtra            246
Delhi                  246
Bihar                  164
Tamil Nadu             164
Meghalaya                0
Uttarakhand              0
Uttar Pradesh            0
Tripura                  0
Telangana                0
Sikkim                   0
Rajasthan                0
Punjab                   0
Puducherry               0
Orissa                   0
Nagaland                 0
Mizoram                  0
Andaman and Nicobar      0
Manipur                  0
Andhra Pradesh           0
Madhya Pradesh           0
Kerala                   0
Karnataka                0
Jharkhand                0
Himachal Pradesh         0
Haryana                  0
Gujarat                  0
Goa                      0
Chhattisgarh             0
Chandigarh               0
Assam                    0
West Bengal              0
Name: state, dtype: int64

We see that while retail data is available for all states (excluding Chhattisgarh, Sikkim, Manipur, as discussed above), wholesale data is restricted to only four states. This is inadequate to conduct a meaningful analysis of wholesale prices.

## Data Analysis

For simplification, the index we use will just be an average of all commoities, without any weights attached.

### Retail Inflation

In [25]:
#Step 1: Creating price baskets for each state per year
df_retail_inflation = df_retail.groupby(['state', 'year'])['inrprice'].mean().reset_index()
df_retail_inflation = df_retail_inflation.rename({'inrprice':'price_basket'}, axis=1)
df_retail_inflation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 279 entries, 0 to 278
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   state         279 non-null    category
 1   year          279 non-null    int64   
 2   price_basket  252 non-null    float64 
dtypes: category(1), float64(1), int64(1)
memory usage: 6.0 KB


In [26]:
df_retail_inflation.head()

state  year  price_basket
0  Andaman and Nicobar  2014     73.253351
1  Andaman and Nicobar  2015     79.574900
2  Andaman and Nicobar  2016     80.114581
3  Andaman and Nicobar  2017     77.084348
4  Andaman and Nicobar  2018     73.682010

In [27]:
#Checking whether the above groupby method yields the correct values
df_test = df_retail[(df_retail['state']=='Andaman and Nicobar') & (df_retail['year']==2014)]
df_test['inrprice'].describe()
#Success! The mean generated by calling describe() is equal to the mean calculated above (see row 1 of above df)

count    191.000000
mean      73.253351
std       56.414716
min       15.070000
25%       35.615000
50%       52.000000
75%       99.250000
max      240.000000
Name: inrprice, dtype: float64

In [28]:
df_retail_inflation['previous_year_price'] = df_retail_inflation['price_basket'].shift(1)
df_retail_inflation.head()

state  year  price_basket  previous_year_price
0  Andaman and Nicobar  2014     73.253351                  NaN
1  Andaman and Nicobar  2015     79.574900            73.253351
2  Andaman and Nicobar  2016     80.114581            79.574900
3  Andaman and Nicobar  2017     77.084348            80.114581
4  Andaman and Nicobar  2018     73.682010            77.084348

In [29]:
#Measuring year on year inflation
df_retail_inflation['yoy_inflation'] = df_retail_inflation['price_basket']-df_retail_inflation['previous_year_price']
df_retail_inflation['yoy_inflation'] = df_retail_inflation['yoy_inflation']/df_retail_inflation['previous_year_price']*100

In [30]:
df_retail_inflation = df_retail_inflation.drop(['previous_year_price'], axis=1)

In [31]:
df_retail_inflation['state'].value_counts()

Andaman and Nicobar    9
Manipur                9
Uttarakhand            9
Uttar Pradesh          9
Tripura                9
Telangana              9
Tamil Nadu             9
Sikkim                 9
Rajasthan              9
Punjab                 9
Puducherry             9
Orissa                 9
Nagaland               9
Mizoram                9
Meghalaya              9
Maharashtra            9
Andhra Pradesh         9
Madhya Pradesh         9
Kerala                 9
Karnataka              9
Jharkhand              9
Himachal Pradesh       9
Haryana                9
Gujarat                9
Goa                    9
Delhi                  9
Chhattisgarh           9
Chandigarh             9
Bihar                  9
Assam                  9
West Bengal            9
Name: state, dtype: int64

In [32]:
df_retail_inflation.head(10)

state  year  price_basket  yoy_inflation
0  Andaman and Nicobar  2014     73.253351            NaN
1  Andaman and Nicobar  2015     79.574900       8.629707
2  Andaman and Nicobar  2016     80.114581       0.678206
3  Andaman and Nicobar  2017     77.084348      -3.782375
4  Andaman and Nicobar  2018     73.682010      -4.413786
5  Andaman and Nicobar  2019     79.304306       7.630487
6  Andaman and Nicobar  2020     92.743786      16.946722
7  Andaman and Nicobar  2021    103.519931      11.619264
8  Andaman and Nicobar  2022    106.708605       3.080251
9       Andhra Pradesh  2014     55.048848     -48.411988

A problem we  encounter here is the prices of different states being compared together. For example, in row 9, the Andhra Pradesh 2014 price has been compared with the Andaman and Nicobar price of 2022, which gives us a redundant value for inflation in Andhra Pradesh in 2014. We need to drop all such values.

In [33]:
#Substitute all 2014 inflation values with null values
df_retail_inflation.loc[df_retail_inflation['year']==2014, 'yoy_inflation'] = np.nan
df_retail_inflation[df_retail_inflation['year']==2014]['yoy_inflation'].describe()

count    0.0
mean     NaN
std      NaN
min      NaN
25%      NaN
50%      NaN
75%      NaN
max      NaN
Name: yoy_inflation, dtype: float64

In [34]:
#Calculating total inflation from 2014-2022
df_retail_inflation['2014_price'] = df_retail_inflation['price_basket'].shift(8)
df_retail_inflation['total_inflation'] = df_retail_inflation['price_basket']-df_retail_inflation['2014_price']
df_retail_inflation['total_inflation'] = df_retail_inflation['total_inflation']/df_retail_inflation['2014_price']*100
df_retail_inflation.head(10)

state  year  price_basket  yoy_inflation  2014_price  \
0  Andaman and Nicobar  2014     73.253351            NaN         NaN   
1  Andaman and Nicobar  2015     79.574900       8.629707         NaN   
2  Andaman and Nicobar  2016     80.114581       0.678206         NaN   
3  Andaman and Nicobar  2017     77.084348      -3.782375         NaN   
4  Andaman and Nicobar  2018     73.682010      -4.413786         NaN   
5  Andaman and Nicobar  2019     79.304306       7.630487         NaN   
6  Andaman and Nicobar  2020     92.743786      16.946722         NaN   
7  Andaman and Nicobar  2021    103.519931      11.619264         NaN   
8  Andaman and Nicobar  2022    106.708605       3.080251   73.253351   
9       Andhra Pradesh  2014     55.048848            NaN   79.574900   

   total_inflation  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
5              NaN  
6              NaN  
7              NaN  
8        45.670612  
9       -30.821342

In [35]:
df_retail_inflation = df_retail_inflation.drop(['2014_price'], axis=1)

In [36]:
#Dropping redundant values and rounding all values
df_retail_inflation.loc[df_retail_inflation['year']!=2022, 'total_inflation'] = np.nan
df_retail_inflation['price_basket'] = df_retail_inflation['price_basket'].round(2)
df_retail_inflation['yoy_inflation'] = df_retail_inflation['yoy_inflation'].round(2)
df_retail_inflation['total_inflation'] = df_retail_inflation['total_inflation'].round(2)
df_retail_inflation.head(18)

state  year  price_basket  yoy_inflation  total_inflation
0   Andaman and Nicobar  2014         73.25            NaN              NaN
1   Andaman and Nicobar  2015         79.57           8.63              NaN
2   Andaman and Nicobar  2016         80.11           0.68              NaN
3   Andaman and Nicobar  2017         77.08          -3.78              NaN
4   Andaman and Nicobar  2018         73.68          -4.41              NaN
5   Andaman and Nicobar  2019         79.30           7.63              NaN
6   Andaman and Nicobar  2020         92.74          16.95              NaN
7   Andaman and Nicobar  2021        103.52          11.62              NaN
8   Andaman and Nicobar  2022        106.71           3.08            45.67
9        Andhra Pradesh  2014         55.05            NaN              NaN
10       Andhra Pradesh  2015         61.93          12.50              NaN
11       Andhra Pradesh  2016         67.18           8.47              NaN
12       Andhra Pradesh  2017         65.17          -2.99              NaN
13       Andhra Pradesh  2018         62.01          -4.84              NaN
14       Andhra Pradesh  2019         66.98           8.01              NaN
15       Andhra Pradesh  2020         77.16          15.20              NaN
16       Andhra Pradesh  2021         92.79          20.25              NaN
17       Andhra Pradesh  2022        100.94           8.78            83.36

In [37]:
#Sorting from lowest to highest inflation
df_retail_inflation.sort_values(by='total_inflation', ascending=True).head()

state  year  price_basket  yoy_inflation  total_inflation
8    Andaman and Nicobar  2022        106.71           3.08            45.67
170              Mizoram  2022        107.47          12.66            45.80
107            Jharkhand  2022         92.33           5.99            49.86
269          Uttarakhand  2022         94.54          11.24            56.04
44            Chandigarh  2022         93.95           7.20            56.28

In [38]:
#Sorting from highest to lowest inflation
df_retail_inflation.sort_values(by='total_inflation', ascending=False).head()

state  year  price_basket  yoy_inflation  total_inflation
26            Assam  2022         90.58           4.41           116.30
251         Tripura  2022         95.27           6.19            90.72
215       Rajasthan  2022         93.84           6.25            87.31
17   Andhra Pradesh  2022        100.94           8.78            83.36
71              Goa  2022        104.35           7.03            79.52

We have succesfully conducted the first piece of this analysis! As per the data, Assam had the highest total retail food inflation (116.3%) in the 2014-22 period, while Andaman and Nicobar Islands had the lowest (45.67%).

### Average retail prices over time

In [39]:
df_retail_avg = df_retail.groupby('state')['inrprice'].mean().reset_index()
df_retail_avg #Keep this, useful to compare aggregate prices across states over time
df_retail_avg = df_retail_avg.rename({'inrprice':'avginrprice'}, axis=1) #renaming inrprice column
df_retail_avg['avginrprice'] = df_retail_avg['avginrprice'].round(2)
df_retail_avg.head()

state  avginrprice
0  Andaman and Nicobar        87.66
1       Andhra Pradesh        80.17
2                Assam        66.11
3                Bihar        79.27
4           Chandigarh        69.91

In [40]:
df_retail_avg.sort_values(by='avginrprice', ascending=True).head()

state  avginrprice
2        Assam        66.11
20      Orissa        68.97
4   Chandigarh        69.91
27     Tripura        70.00
9      Haryana        70.91

In [41]:
df_retail_avg.sort_values(by='avginrprice', ascending=False).head()

state  avginrprice
0   Andaman and Nicobar        87.66
18              Mizoram        86.61
12            Karnataka        85.72
17            Meghalaya        83.23
7                   Goa        82.68

Our analysis now tells us that the highest average prices in the 2014-22 period were in the Andaman and Nicobar Islands, while the lowest were in Assam. This is an interesting result, considering we observed earlier that the Andaman and Nicobar Islands had the *lowest* inflation rate and Assam the *highest*. One can probably attribute these inflation rates to base rate effects.

### Commodity-wise prices

Now, on to the next section of the analysis. We shift our focus from state-wise analysis to commodity-wise analysis. Here, we answer the question: Which categories and commodities had the highest/lowest rates of retail inflation in the given time period?

In [42]:
df_retail.head()

state         district      market            category  \
29336  Andaman and Nicobar  Andaman Islands  Port Blair  cereals and tubers   
29337  Andaman and Nicobar  Andaman Islands  Port Blair  cereals and tubers   
29338  Andaman and Nicobar  Andaman Islands  Port Blair  cereals and tubers   
29339  Andaman and Nicobar  Andaman Islands  Port Blair  cereals and tubers   
29340  Andaman and Nicobar  Andaman Islands  Port Blair      milk and dairy   

                commodity unit pricetype  inrprice  year  
29336            Potatoes   KG    Retail      26.0  2014  
29337                Rice   KG    Retail      52.0  2014  
29338               Wheat   KG    Retail      27.0  2014  
29339         Wheat flour   KG    Retail      44.0  2014  
29340  Milk (pasteurized)    L    Retail      52.0  2014

In [43]:
#Grouping food category prices by year
df_category = df_retail.groupby(['category', 'year'])['inrprice'].mean().reset_index()
df_category = df_category.rename({'inrprice':'avg_national_price'}, axis=1)
print(df_category.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   category            54 non-null     category
 1   year                54 non-null     int64   
 2   avg_national_price  54 non-null     float64 
dtypes: category(1), float64(1), int64(1)
memory usage: 1.2 KB
None


In [44]:
#Checking for uniformity in category distribution
df_category['category'].value_counts()

cereals and tubers       9
milk and dairy           9
miscellaneous food       9
oil and fats             9
pulses and nuts          9
vegetables and fruits    9
Name: category, dtype: int64

In [45]:
#Calculating y-o-y inflation similar to our state-wise analysis:
df_category['prev_year_price'] = df_category['avg_national_price'].shift(1)
df_category['yoy_inflation'] = df_category['avg_national_price']-df_category['prev_year_price']
df_category['yoy_inflation'] = df_category['yoy_inflation']/df_category['prev_year_price']*100
df_category.loc[df_category['year']==2014, 'yoy_inflation'] = np.nan #Removing redundant values
df_category.head(10)

category  year  avg_national_price  prev_year_price  \
0  cereals and tubers  2014           23.955859              NaN   
1  cereals and tubers  2015           21.943914        23.955859   
2  cereals and tubers  2016           23.624837        21.943914   
3  cereals and tubers  2017           23.773876        23.624837   
4  cereals and tubers  2018           25.322839        23.773876   
5  cereals and tubers  2019           26.909035        25.322839   
6  cereals and tubers  2020           31.309094        26.909035   
7  cereals and tubers  2021           29.122678        31.309094   
8  cereals and tubers  2022           30.136622        29.122678   
9      milk and dairy  2014           36.866884        30.136622   

   yoy_inflation  
0            NaN  
1      -8.398551  
2       7.660088  
3       0.630856  
4       6.515400  
5       6.263895  
6      16.351607  
7      -6.983326  
8       3.481631  
9            NaN

In [46]:
df_category = df_category.drop(['prev_year_price'], axis=1)

In [47]:
#Total inflation
df_category['2014_price'] = df_category['avg_national_price'].shift(8)
df_category.loc[df_category['year']!=2022, '2014_price'] = np.nan
df_category['total_inflation'] = df_category['avg_national_price']-df_category['2014_price']
df_category['total_inflation'] = df_category['total_inflation']/df_category['2014_price']*100
df_category.head(10)

category  year  avg_national_price  yoy_inflation  2014_price  \
0  cereals and tubers  2014           23.955859            NaN         NaN   
1  cereals and tubers  2015           21.943914      -8.398551         NaN   
2  cereals and tubers  2016           23.624837       7.660088         NaN   
3  cereals and tubers  2017           23.773876       0.630856         NaN   
4  cereals and tubers  2018           25.322839       6.515400         NaN   
5  cereals and tubers  2019           26.909035       6.263895         NaN   
6  cereals and tubers  2020           31.309094      16.351607         NaN   
7  cereals and tubers  2021           29.122678      -6.983326         NaN   
8  cereals and tubers  2022           30.136622       3.481631   23.955859   
9      milk and dairy  2014           36.866884            NaN         NaN   

   total_inflation  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
5              NaN  
6              NaN  
7              NaN  
8        25.800634  
9              NaN

In [48]:
#Cleaning up
df_category = df_category.drop(['2014_price'], axis=1)
df_category['avg_national_price'] = df_category['avg_national_price'].round(2)
df_category['yoy_inflation'] = df_category['yoy_inflation'].round(2)
df_category['total_inflation'] = df_category['total_inflation'].round(2)

In [49]:
df_category

category  year  avg_national_price  yoy_inflation  \
0      cereals and tubers  2014               23.96            NaN   
1      cereals and tubers  2015               21.94          -8.40   
2      cereals and tubers  2016               23.62           7.66   
3      cereals and tubers  2017               23.77           0.63   
4      cereals and tubers  2018               25.32           6.52   
5      cereals and tubers  2019               26.91           6.26   
6      cereals and tubers  2020               31.31          16.35   
7      cereals and tubers  2021               29.12          -6.98   
8      cereals and tubers  2022               30.14           3.48   
9          milk and dairy  2014               36.87            NaN   
10         milk and dairy  2015               39.57           7.33   
11         milk and dairy  2016               40.39           2.07   
12         milk and dairy  2017               42.15           4.37   
13         milk and dairy  2018               43.15           2.37   
14         milk and dairy  2019               44.84           3.91   
15         milk and dairy  2020               46.97           4.74   
16         milk and dairy  2021               49.06           4.46   
17         milk and dairy  2022               50.82           3.58   
18     miscellaneous food  2014               84.45            NaN   
19     miscellaneous food  2015               75.80         -10.25   
20     miscellaneous food  2016               73.81          -2.61   
21     miscellaneous food  2017               77.23           4.63   
22     miscellaneous food  2018               77.29           0.08   
23     miscellaneous food  2019               77.81           0.67   
24     miscellaneous food  2020               81.80           5.13   
25     miscellaneous food  2021               94.86          15.96   
26     miscellaneous food  2022               96.34           1.56   
27           oil and fats  2014               97.55            NaN   
28           oil and fats  2015               93.33          -4.33   
29           oil and fats  2016               94.56           1.32   
30           oil and fats  2017               95.42           0.91   
31           oil and fats  2018               97.63           2.32   
32           oil and fats  2019              100.37           2.80   
33           oil and fats  2020              114.18          13.76   
34           oil and fats  2021              151.72          32.88   
35           oil and fats  2022              168.25          10.89   
36        pulses and nuts  2014               56.32            NaN   
37        pulses and nuts  2015               85.22          51.32   
38        pulses and nuts  2016              100.44          17.86   
39        pulses and nuts  2017               79.55         -20.80   
40        pulses and nuts  2018               67.12         -15.62   
41        pulses and nuts  2019               77.03          14.77   
42        pulses and nuts  2020               94.29          22.41   
43        pulses and nuts  2021              101.66           7.81   
44        pulses and nuts  2022              101.50          -0.16   
45  vegetables and fruits  2014               24.77            NaN   
46  vegetables and fruits  2015               29.72          20.00   
47  vegetables and fruits  2016               22.35         -24.81   
48  vegetables and fruits  2017               28.65          28.23   
49  vegetables and fruits  2018               23.76         -17.07   
50  vegetables and fruits  2019               39.33          65.50   
51  vegetables and fruits  2020               36.84          -6.32   
52  vegetables and fruits  2021               31.76         -13.80   
53  vegetables and fruits  2022               31.82           0.19   

    total_inflation  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN  
5               NaN  
6           

In [50]:
df_category.sort_values(by='total_inflation', ascending=True).head()

category  year  avg_national_price  yoy_inflation  \
26     miscellaneous food  2022               96.34           1.56   
8      cereals and tubers  2022               30.14           3.48   
53  vegetables and fruits  2022               31.82           0.19   
17         milk and dairy  2022               50.82           3.58   
35           oil and fats  2022              168.25          10.89   

    total_inflation  
26            14.07  
8             25.80  
53            28.48  
17            37.85  
35            72.47

In [51]:
df_category.sort_values(by='total_inflation', ascending=False).head()

category  year  avg_national_price  yoy_inflation  \
44        pulses and nuts  2022              101.50          -0.16   
35           oil and fats  2022              168.25          10.89   
17         milk and dairy  2022               50.82           3.58   
53  vegetables and fruits  2022               31.82           0.19   
8      cereals and tubers  2022               30.14           3.48   

    total_inflation  
44            80.22  
35            72.47  
17            37.85  
53            28.48  
8             25.80

We see that the highest price rise in the 2014-22 period was witnessed by pulses and nuts (80.22%), while the lowest was for the category miscellaneous food (14.07%).

### The post-pandemic inflation

Let us also use the wrangled dataset for insights on retail inflation as per states and food categories in the context of the post-pandemic inflation.

In 2020, the government of India imposed a nationwide lockdown to prevent the virus from spreading, and it was only later that year that the lockdown was lifted. The lockdown period was accompanied by deflation in prices, and subsequently, the post-lockdown period was a high-inflation one, owing to the rebound of economic activity and demand. This inflation has been persistent till date (August 2023), and so it would be pertinent to measure food retail inflation in the years 2020-22 (the latest year till which we have data).

The method for doing this will be the exact same as our above two analyses for state-wise and category-wise inflation.

In [52]:
#State-wise analysis
df_post_covid = df_retail_inflation[df_retail_inflation['year']>=2020]
df_post_covid.head(10)

state  year  price_basket  yoy_inflation  total_inflation
6   Andaman and Nicobar  2020         92.74          16.95              NaN
7   Andaman and Nicobar  2021        103.52          11.62              NaN
8   Andaman and Nicobar  2022        106.71           3.08            45.67
15       Andhra Pradesh  2020         77.16          15.20              NaN
16       Andhra Pradesh  2021         92.79          20.25              NaN
17       Andhra Pradesh  2022        100.94           8.78            83.36
24                Assam  2020         72.85          12.86              NaN
25                Assam  2021         86.75          19.08              NaN
26                Assam  2022         90.58           4.41           116.30
33                Bihar  2020         75.02          20.77              NaN

In [53]:
#Redefining total_inflation
df_post_covid['2020_price'] = df_post_covid['price_basket'].shift(2)
df_post_covid['total_inflation'] = df_post_covid['price_basket']-df_post_covid['2020_price']
df_post_covid['total_inflation'] = df_post_covid['total_inflation']/df_post_covid['2020_price']*100
df_post_covid.head()

/var/folders/sx/wtgz65jx1vqg_v5k12yf0wfr0000gn/T/ipykernel_74588/2601180568.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_post_covid['2020_price'] = df_post_covid['price_basket'].shift(2)
/var/folders/sx/wtgz65jx1vqg_v5k12yf0wfr0000gn/T/ipykernel_74588/2601180568.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_post_covid['total_inflation'] = df_post_covid['price_basket']-df_post_covid['2020_price']
/var/folders/sx/wtgz65jx1vqg_v5k12yf0wfr0000gn/T/ipykernel_74588/2601180568.py:4: SettingWithC

state  year  price_basket  yoy_inflation  total_inflation  \
6   Andaman and Nicobar  2020         92.74          16.95              NaN   
7   Andaman and Nicobar  2021        103.52          11.62              NaN   
8   Andaman and Nicobar  2022        106.71           3.08        15.063619   
15       Andhra Pradesh  2020         77.16          15.20       -25.463679   
16       Andhra Pradesh  2021         92.79          20.25       -13.044701   

    2020_price  
6          NaN  
7          NaN  
8        92.74  
15      103.52  
16      106.71

In [54]:
df_post_covid = df_post_covid.drop(['2020_price'], axis=1)

In [55]:
#Cleaning up
df_post_covid.loc[df_post_covid['year']!=2022, 'total_inflation'] = np.nan
df_post_covid['total_inflation'] = df_post_covid['total_inflation'].round(2)
df_post_covid.head(6)

state  year  price_basket  yoy_inflation  total_inflation
6   Andaman and Nicobar  2020         92.74          16.95              NaN
7   Andaman and Nicobar  2021        103.52          11.62              NaN
8   Andaman and Nicobar  2022        106.71           3.08            15.06
15       Andhra Pradesh  2020         77.16          15.20              NaN
16       Andhra Pradesh  2021         92.79          20.25              NaN
17       Andhra Pradesh  2022        100.94           8.78            30.82

In [56]:
df_post_covid.sort_values(by='total_inflation', ascending=True).head()

state  year  price_basket  yoy_inflation  total_inflation
170              Mizoram  2022        107.47          12.66             8.05
107            Jharkhand  2022         92.33           5.99            12.39
8    Andaman and Nicobar  2022        106.71           3.08            15.06
197           Puducherry  2022         95.91           2.59            15.53
71                   Goa  2022        104.35           7.03            17.88

In [57]:
df_post_covid.sort_values(by='total_inflation', ascending=False).head()

state  year  price_basket  yoy_inflation  total_inflation
251      Tripura  2022         95.27           6.19            38.07
80       Gujarat  2022         94.41           7.89            35.37
206       Punjab  2022        101.42          11.25            33.57
143  Maharashtra  2022        100.70           8.56            33.38
44    Chandigarh  2022         93.95           7.20            31.05

In the post-pandemic inflation period, Tripura saw the most severe retail food price rise (38.07%), while Mizoram saw the least (8.05%).

In [58]:
#Category-wise analysis
df_post_covid_category = df_category[df_category['year'] >= 2020]
df_post_covid_category.head()

category  year  avg_national_price  yoy_inflation  \
6   cereals and tubers  2020               31.31          16.35   
7   cereals and tubers  2021               29.12          -6.98   
8   cereals and tubers  2022               30.14           3.48   
15      milk and dairy  2020               46.97           4.74   
16      milk and dairy  2021               49.06           4.46   

    total_inflation  
6               NaN  
7               NaN  
8              25.8  
15              NaN  
16              NaN

In [59]:
#Redefining total inflation
df_post_covid_category['2020_price'] = df_post_covid_category['avg_national_price'].shift(2)
df_post_covid_category.loc[df_post_covid_category['year']!=2022, '2020_price'] = np.nan
df_post_covid_category['total_inflation'] = df_post_covid_category['avg_national_price']-df_post_covid_category['2020_price']
df_post_covid_category['total_inflation'] = 100*df_post_covid_category['total_inflation']/df_post_covid_category['2020_price']
df_post_covid_category['total_inflation'] = df_post_covid_category['total_inflation'].round(2)
df_post_covid_category.head(6)

/var/folders/sx/wtgz65jx1vqg_v5k12yf0wfr0000gn/T/ipykernel_74588/3749562601.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_post_covid_category['2020_price'] = df_post_covid_category['avg_national_price'].shift(2)
/var/folders/sx/wtgz65jx1vqg_v5k12yf0wfr0000gn/T/ipykernel_74588/3749562601.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_post_covid_category['total_inflation'] = df_post_covid_category['avg_national_price']-df_post_covid_category['2020_price']
/var/folders/sx/wtgz65jx1vqg_v5k12yf0

category  year  avg_national_price  yoy_inflation  \
6   cereals and tubers  2020               31.31          16.35   
7   cereals and tubers  2021               29.12          -6.98   
8   cereals and tubers  2022               30.14           3.48   
15      milk and dairy  2020               46.97           4.74   
16      milk and dairy  2021               49.06           4.46   
17      milk and dairy  2022               50.82           3.58   

    total_inflation  2020_price  
6               NaN         NaN  
7               NaN         NaN  
8             -3.74       31.31  
15              NaN         NaN  
16              NaN         NaN  
17             8.20       46.97

In [60]:
df_post_covid_category = df_post_covid_category.drop(['2020_price'], axis=1)

In [61]:
df_post_covid_category.sort_values(by='total_inflation', ascending=True).head()

category  year  avg_national_price  yoy_inflation  \
53  vegetables and fruits  2022               31.82           0.19   
8      cereals and tubers  2022               30.14           3.48   
44        pulses and nuts  2022              101.50          -0.16   
17         milk and dairy  2022               50.82           3.58   
26     miscellaneous food  2022               96.34           1.56   

    total_inflation  
53           -13.63  
8             -3.74  
44             7.65  
17             8.20  
26            17.78

In [62]:
df_post_covid_category.sort_values(by='total_inflation', ascending=False).head()

category  year  avg_national_price  yoy_inflation  \
35        oil and fats  2022              168.25          10.89   
26  miscellaneous food  2022               96.34           1.56   
17      milk and dairy  2022               50.82           3.58   
44     pulses and nuts  2022              101.50          -0.16   
8   cereals and tubers  2022               30.14           3.48   

    total_inflation  
35            47.36  
26            17.78  
17             8.20  
44             7.65  
8             -3.74

Finally, we see that in the post-pandemic period, oil and fats saw the highest price rise (47.36%), while the lowest inflation was that of vegetables and fruits, which was in fact a price fall of -13.63%.